File: 1026. Maximum-Difference-Between-Node-and-Ancestor/Solution1026.cpp
class Solution1026 {
public:
    int maxAncestorDiff(TreeNode* root) {
        return helper(root, root->val, root->val);
    }
    int helper(TreeNode* root, int minVal, int maxVal) {
        if (root == nullptr) {
            return maxVal - minVal;
        }
        minVal = min(minVal, root->val);
        maxVal = max(maxVal, root->val);
        int left = helper(root->left, minVal, maxVal);
        int right = helper(root->right, minVal, maxVal);
        return max(left, right);
    }
};

File: 1026. Maximum-Difference-Between-Node-and-Ancestor/Solution1026.py
class Solution1026:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        return self.helper(root, root.val, root.val)
    
    def helper(self, root: TreeNode, minVal: int, maxVal: int) -> int:
        if not root:
            return maxVal - minVal
        maxVal = max(maxVal, root.val)
        minVal = min(minVal, root.val)
        left = self.helper(root.left, minVal, maxVal)
        right = self.helper(root.right, minVal, maxVal)
        return max(left, right)